### ISSUE TITLE
Helm : Add the ArgoCD annotation on some resources to ignore change

### ISSUE THREAD

**Issue by pbrissaud** [linkerd/linkerd2#8440]
### What problem are you trying to solve?

I successfully deployed the Linkerd2 helm chart with ArgoCD (thanks for the good tutorial) ! But everyday when the secret is rotating, the application become **Out-Of-Sync**  because some Kubernetes secrets change.

### How should the problem be solved?

I suggest adding this annotation `argocd.argoproj.io/compare-options: IgnoreExtraneous` on the resources which can change. For the moment, I detect these resources  : 
- Secret : linkerd-policy-validator-k8s-tls
- Secret : linkerd-proxy-injector-k8s-tls
- Secret : linkerd-sp-validator-k8s-tls

### Any alternatives you've considered?

For people not using ArgoCD, these annotations don't change anything 

### Would you like to work on this feature?

yes

**Comment by olix0r** [comment]
> But everyday when the secret is rotating

Linkerd's not doing that though, right? What is rotating the secret? Cert-manager? Can we configure whatever is rotating the secret to set these annotations?

In general, we're reluctant to add arbitrary labels in Linkerd to account for specific integrations like argo cd...

**Comment by pbrissaud** [comment]
In the documentation, it is indicated on this [page](https://linkerd.io/2.11/tasks/automatically-rotating-control-plane-tls-credentials/) : 
> While Linkerd automatically rotates the TLS certificates for data plane proxies every 24 hours

And it is true because I synced the App yesterday and today it was out of sync: 
<img width="346" alt="image" src="https://user-images.githubusercontent.com/27202456/167161955-b635324c-d1e4-4aa7-b179-bfdc7c5f5f00.png">

Yeah I use cert-manager for the the control plane as it described on the doc : https://linkerd.io/2.11/tasks/automatically-rotating-control-plane-tls-credentials

> In general, we're reluctant to add arbitrary labels in Linkerd to account for specific integrations like argo cd...

I can understand but as ArgoCD is presented like an official solution to install Linkerd, it would be nice to add these annotation

**Comment by olix0r** [comment]
> While Linkerd automatically rotates the TLS certificates for data plane proxies every 24 hours

These proxy's credentials are ephemeral and aren't related to the *k8s-tls secrets. The secrets you reference are used by Linkerd's admission controllers to run HTTPS servers that are called by the Kubernetes API server.

> Yeah I use cert-manager for the the control plane

Unless I'm mistaken, Linkerd isn't actually creating any of these resources for you--cert-manager is. So there's nothing we can change in Linkerd to solve this problem; but I imagine others have encountered this when using cert-manager and Argo CD. We should find out what the recommended pattern is for cert-manager so we can update our documentation.

**Comment by pbrissaud** [comment]
> These proxy credentials are ephemeral and aren't related to the *k8s-tls secrets

Ok my bad then. Thanks for the information 

> Linkerd isn't actually creating any of these resources for you--cert-manager is

Nope cert manager only creates the identity issuer certificate (as a Kubernetes secret). These 3 secrets are created with Helm in the `destination-tabac.yml` template file.

Will you be interested to see my Argo Application resource ?

**Comment by olix0r** [comment]
> Nope cert manager only creates the identity issuer certificate (as a Kubernetes secret). These 3 secrets are created with Helm in the destination-rbac.yml template file.

Interesting. I'm a little confused about how this works, then. Can you share your values.yaml or the output of `kubectl get cm -n linkerd linkerd-config -o yaml`?

Specifically the doc at https://linkerd.io/2.11/tasks/automatically-rotating-webhook-tls-credentials/ outlines how to configure cert-manager to manage these k8s-tls secrets. That doc outlines that Linkerd should be installed with the values overrides:
```
  --set policyValidator.externalSecret=true \
  --set proxyInjector.externalSecret=true \
  --set profileValidator.externalSecret=true \
```

And these overrides disable Linkerd from generating these secrets:
https://github.com/linkerd/linkerd2/blob/c1a1430d1a524ff6a74ebf3a5127d276b5a3b26a/charts/linkerd-control-plane/templates/destination-rbac.yaml#L59-L75

Basically, if these secrets are installed by Linkerd, they shouldn't change unless you reinstall Linkerd. And then the argo annotation isn't necessary. Otherwise, these secrets are managed by cert-manager, and then changing Linkerd's manifests won't help anything.

If you can share more details about how you configured Linkerd/Cert-manager, we may be able to get to the bottom of it :)

**Comment by pbrissaud** [comment]
Here's my value file : 
```yaml
identity:
  issuer:
    scheme: "kubernetes.io/tls"
installNamespace: false
identityTrustAnchorsPEM: |
  -----BEGIN CERTIFICATE-----
  ......
  ---END CERTIFICATE----
```

> That doc outlines that Linkerd should be installed with the values overrides :

Effectively, I don't add theses values, I will try to add them. Maybe theses values should be added to this documentation https://linkerd.io/2.11/tasks/automatically-rotating-control-plane-tls-credentials as the situation is quite the same.

**Comment by olix0r** [comment]
>  Maybe theses values should be added to this documentation https://linkerd.io/2.11/tasks/automatically-rotating-control-plane-tls-credentials as the situation is quite the same.

This document only deals with managing the `linkerd-identity-issuer` Secret and the only cert-manager integration it covers is:
```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: linkerd-identity-issuer
  namespace: linkerd
spec:
  secretName: linkerd-identity-issuer
  ...
```

The other document covers creating [cert-manager configurations for the other *k8s-tls secrets](https://linkerd.io/2.11/tasks/automatically-rotating-webhook-tls-credentials/#issuing-certificates-and-writing-them-to-secrets). For example:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: linkerd-policy-validator
  namespace: linkerd
spec:
  secretName: linkerd-policy-validator-k8s-tls
  ...
```

Presumably, you have these `Certificate` resources in the linkerd namespace as well?

I know this is all a bit complicated. It sounds like we could use a better overview of all of the different cert management pieces. But the two docs we're discussing cover configuring different credentials:

* https://linkerd.io/2.11/tasks/automatically-rotating-control-plane-tls-credentials describes how to setup the issuer credentials that are used by Linkerd's `identity` controller to automatically provision ephemeral proxy certificates.
* https://linkerd.io/2.11/tasks/automatically-rotating-webhook-tls-credentials describes how to setup the controller web hook certificates (*-k8s-tls)

**Comment by pbrissaud** [comment]
I haven't had time to read the documentation about webhooks certificates so far so I think the time has come ! 

Maybe this part should be added optionally here : https://linkerd.io/2.11/tasks/gitops/

Thanks again for your help !

**Comment by pbrissaud** [comment]
I follow the doc, everything is working fine now ! 

I think this issue can be closed now ?
